Source/utils/language.cpp

```diff
@@ -0,0 +1,129 @@
+#include "options.h"
+#include "utils/paths.h"
+#include <map>
+
+using namespace devilution;
+#define MO_MAGIC 0x950412de
+
+namespace {
+
+struct cstring_cmp {
+	bool operator()(const char *s1, const char *s2) const
+	{
+		return strcmp(s1, s2) < 0;
+	}
+};
+
+std::map<const char *, const char *, cstring_cmp> map;
+
+struct mo_head {
+	uint32_t magic;
+	struct {
+		uint16_t major;
+		uint16_t minor;
+	} revision;
+
+	uint32_t nb_mappings;
+	uint32_t src_offset;
+	uint32_t dst_offset;
+};
+
+struct mo_entry {
+	uint32_t length;
+	uint32_t offset;
+};
+
+char *read_entry(FILE *fp, mo_entry *e)
+{
+	void *data;
+
+	if (fseek(fp, e->offset, SEEK_SET)) {
+		return nullptr;
+	}
+
+	if (!(data = calloc(e->length, sizeof(char)))) {
+		return nullptr;
+	}
+
+	if (fread(data, sizeof(char), e->length, fp) != e->length) {
+		free(data);
+		return nullptr;
+	}
+
+	return static_cast<char *>(data);
+}
+} // namespace
+
+const char *LanguageTranslate(const char *key)
+{
+	auto it = map.find(key);
+	if (it == map.end()) {
+		return key;
+	}
+
+	return it->second;
+}
+
+void LanguageInitialize()
+{
+	auto path = GetLangPath() + "./" + sgOptions.Language.szCode + ".gmo";
```

Comment:
why does it look for files with .gmo instead of .mo?

Reply:
The CMake wrapper around gettext gives them this name. The g in .gmo stands for GNU, and is briefly mentioned at https://www.gnu.org/software/gettext/manual/html_node/Files.html

As far as I understand it, a GMO file is essentially the same thing as a MO file, but they are produced by GNU gettext. I have not yet figured out if there are practical differences. In any case, the parser was implemented based on the description from https://www.gnu.org/software/gettext/manual/html_node/MO-Files.html, which I assume reflects GMO files correctly.